### PR TITLE
Add Fold Conv2D+BatchNormal

### PR DESCRIPTION
**Pull request by zhen8838**
- [x] Add fold conv2D+BatchNorm (first convert Relu to Clamp, then combine Clamp with Binary, finally fold Conv2D+BN)
   - [x] Add ReluToClamp,Relu6ToClamp
   - [x] Add CombineClampMul/CombineClampAdd
   - [x] add FoldConv2DMulAdd
- [x] Add dumpFlags parameter in DumpScope ctor, Now the Dumpflag of the subdumper can be modified.
```csharp
        // Can custom dump information in the tests.
        using (var _ = new DumpScope($"Eval", DumpFlags.Evaluator)){
            CompilerServices.Evaluate(expr);
        }
```

I suggest to name CombineRule according to the order of Call. For example:
`Transpose(Pad(x)) => Pad(Transpose(x))`, should be named `CombineTransposePad`.

### CODE REVIEW

**Review by sunnycase — CHANGES_REQUESTED** [review]
Disabled dumps should not be reenable in sub dumppers. Sub dumppers' dump flags should be subset of parent's dump flags.

File: src/Nncase.Tests/Rewrite/RewriteBase.cs [hunk 1]
```diff
@@ -40,6 +40,13 @@ public interface IRewriteCase
     /// Gets the eval inputs dict.
     /// </summary>
     Dictionary<Var, IValue> FeedDict { get; }
+
+    /// <summary>
+    /// check rewrite post callback.
+    /// </summary>
+    /// <param name="post"></param>
+    /// <returns></returns>
+    bool CheckPostCallBack(Function post) => true;
```
Comment (sunnycase) [inline src/Nncase.Tests/Rewrite/RewriteBase.cs:49]: What's the usage of `CheckPostCallBack`?

File: src/Nncase.EGraph/Transform/EGraphPrinter.cs [hunk 2]
```diff
@@ -208,7 +208,11 @@ public DotGraph ConvertEGraphAsDot()
     /// <returns>this dot graph.</returns>
     public DotGraph SaveToStream(Stream output)
     {
-        DotGraph.Build(new StreamWriter(output, leaveOpen: true));
+        using (var writer = new StreamWriter(output))
```
Comment (sunnycase) [inline src/Nncase.EGraph/Transform/EGraphPrinter.cs:211]: Should not close the stream here. If users want the stream be leave open, they can't control the behavior of this method. If users want the stream be disposed, they can use `using` at the caller.

File: src/Nncase.Core/Diagnostics/IDumpper.cs [hunk 3]
```diff
@@ -32,8 +32,9 @@ public interface IDumpper
     /// Create sub dummper.
     /// </summary>
     /// <param name="subDirectory">Sub directory.</param>
+    /// <param name="dumpFlags">Sub dumpFlags.</param>
     /// <returns>Sub dummper.</returns>
-    IDumpper CreateSubDummper(string subDirectory);
+    IDumpper CreateSubDummper(string subDirectory, DumpFlags? dumpFlags);
```
Comment (sunnycase) [inline src/Nncase.Core/Diagnostics/IDumpper.cs:37]: Use default parameter value.

File: src/Nncase.Transform/Rules/Neutral/CombineBinary.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+/// <summary>
+/// clamp(add(x,addConst),min,max) => add(clamp(x,min-addConst,max-addConst),addConst).
+/// </summary>
+[RuleGenerator]
+public sealed partial class CombineClampAdd : IRewriteRule
+{
+    /// <inheritdoc/>
+    public IPattern Pattern { get; init; } = IsClamp(
+      CombineClampUtility.GetBinaryPattern(BinaryOp.Add, "add"),
+      IsTensorConst("min", t => t.Value.ElementType == DataTypes.Float32),
```
Comment (sunnycase) [inline src/Nncase.Transform/Rules/Neutral/CombineBinary.cs:30]: Why constraint the data type to float32?

File: src/Nncase.Tests/Rewrite/RewriteBase.cs (hunk 1 above)
Comment (zhen8838) [reply src/Nncase.Tests/Rewrite/RewriteBase.cs:49]: Sometimes need check the post function structure.

**Comment by zhen8838** [comment]
```csharp
```

File: src/Nncase.EGraph/Transform/EGraphExtractor.cs [hunk 5]
```diff
@@ -129,13 +129,13 @@ private void Visit(EClass eclass)
         while (stack.Any())
         {
             (eclass, var minCostEnode) = stack.Peek();
-            if (_eclassMemo.TryGetValue(eclass, out var expr))
+            if (_eclassMemo.TryGetValue(eclass, out _))
```
Comment (sunnycase) [inline src/Nncase.EGraph/Transform/EGraphExtractor.cs:132]: Use `ContainsKey` instead.

File: src/Nncase.Tests/Rewrite/RewriteBase.cs (hunk 1 above)
Comment (sunnycase) [reply src/Nncase.Tests/Rewrite/RewriteBase.cs:49]: Should naming `ChecksPostCallBack`.

File: src/Nncase.Tests/Rewrite/RewriteBase.cs [hunk 6]
```diff
@@ -833,3 +824,138 @@ public Function PreExpr
         { _input_rhs, Normal(DataTypes.Float32, 0, 1, 1, _input_rhs.CheckedShape.ToValueArray()).Evaluate() },
     };
 }
+
+public sealed class CombineClampAddMul : IRewriteCase
+{
+    private const int _channels = 4; // 256
+    private const int _featrueMap = 3; // 56
+    private readonly Var _input_lhs;
```
Comment (sunnycase) [inline src/Nncase.Tests/Rewrite/RewriteBase.cs:832]: Use camel naming.

File: src/Nncase.Tests/Rewrite/RewriteBase.cs [hunk 7]
```diff
@@ -833,3 +824,138 @@ public Function PreExpr
...
+    {
+        _input_lhs = new Var("_input_lhs", new TensorType(DataTypes.Float32, new[] { 1, _featrueMap, _featrueMap, _channels }));
+        _input_rhs = new Var("_input_rhs", new TensorType(DataTypes.Float32, new[] { 1, _featrueMap, _featrueMap, _channels }));
+    }
+
+    public Function PreExpr
+    {
+        get
+        {
+            var v_0 = _input_lhs + _input_rhs; // [1,_featrueMap,_featrueMap,_channels]
```
Comment (sunnycase) [inline src/Nncase.Tests/Rewrite/RewriteBase.cs:845]: Don't use `_`.

File: src/Nncase.Tests/Rewrite/UnitTestDataFlowRewriteFactory.cs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    [MemberData(nameof(DataOne))]
+    public Task RunOneAsync(IRewriteCase @case) => RunCoreAsync(@case);
+
+    [Theory]
+    [MemberData(nameof(DataAll))]
+    public Task RunAllAsync(IRewriteCase @case) => RunCoreAsync(@case);
+
+    private async Task RunCoreAsync(IRewriteCase @case)
+    {
+        var _ = new DumpScope($"../{@case.Name}", DumpFlags.None);
```
Comment (sunnycase) [inline src/Nncase.Tests/Rewrite/UnitTestDataFlowRewriteFactory.cs:46]: Always using a `DumpScope`.

**Review by sunnycase — APPROVED** [review]

**Pull request merged**